WIP: #1607 - Cleanup selection/cursor API

rayat-amperity opened the pull request:
<!-- ❤️ Thanks for filing a Pull Request on Calva! You are contributing to a better Clojure coding experience. ❤️
## What you can expect:

Here are some things we consider before we merge:

- We make sure the PR is directed at the `dev` branch (unless reasons).
- We figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and will help you test there if it is hard for you to do so. (We appreciate a lot if you take on the work do this of course.)
- We read the source changes. (Surprise! 😄)
- We given feedback and guidance on source changes, if needed. Far from everything is captured in our [code guidelines](https://github.com/BetterThanTomorrow/calva/wiki/Coding-Style).
- We use our domain knowledge to try catch if you have missed some facility already provided in the code base.
- We read the updates to the documentation and help with feedback, trying to keep the documentation site serving well.
- We often check out your code changes and test them.
- We sometimes send the VSIX built from the PR out in the `#calva` channel on slack for others to test. (Actually, we will probably encourage you to do this.)
- We sometimes have a chat within the team about particular changes.
- NB: We also consider if your changes belong in the Calva product we want to maintain. Before you spend a lot of work on a PR, please consider chatting us up first, and filing issues.

We try to be speedy and attentive. Please don't hesitate to bump a PR, or contact us, if we seem to have dropped the ball (that has happened).

We use checklists in order to not forget about important lessons we and others have learnt along the way.

-->

## What has Changed?
<!-- Introduce the change(s) briefly here. Consider explaining why a particular change was implemented the way it was. If you have considered alternative ways to introduce the change, please elaborate a bit on that as well. -->

- The `.selectionLeft` and `.selectionRight` props on EditableDocument and its children have been removed in favor of more basic or explicit calls to `.selection.anchor` and `.selection.active`, respectively.
- All call sites or references to those previous properties have accordingly been modified to reference the suggested replacements.
- All default arguments in functions also have been refactored accordingly.

<!-- Tell us what Github issue(s) your PR is fixing. Consider creating the issue if there isn't one already. -->

Fixes #

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [❓] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing. **Do I need to do this one for an internal change?**
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR. **I think? I tested on mac + win, and no changes were observed in smoke tests. Of course, unit tests have been fixed to pass in light of changes.**
- [❓] Added to or updated docs in this branch, if appropriate. **I will defer to maintainers as to whether more are needed from these changes.**
- [x] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. **I did personally on windows, not sure about maintainers. Mac was "smoke tested" purely in extension host.**
- [X] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [X] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [X] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [X] Created the issue I am fixing/addressing, if it was not present.
- [X] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe

riotrah committed:
fixes #1607 selection api cleanup

PEZ commented on src/paredit/extension.ts:
```diff
@@ -1,19 +1,19 @@
 import * as vscode from 'vscode';
 import {
   commands,
-  window,
+  ConfigurationChangeEvent,
   Event,
   EventEmitter,
   ExtensionContext,
+  window,
   workspace,
-  ConfigurationChangeEvent,
 } from 'vscode';
+import { EditableDocument } from '../cursor-doc/model';
 import * as paredit from '../cursor-doc/paredit';
 import * as docMirror from '../doc-mirror/index';
-import { EditableDocument } from '../cursor-doc/model';
 import { assertIsDefined } from '../utilities';
+import { StatusBar } from './statusbar';
```
Do you have something that autocleans the imports? If so, please disable it on the Calva codebase. We welcome code quality updates in separate PRs (after a discussion, at least). When changes like these come together with some PR with some other topic, they slow the reviews down too much, imo. They also create potential merge conflicts of a slightly unnecessary nature.

riotrah replied:
Yes, that is indeed the case. I'll undo - sorry!

PEZ reviewed (changes requested):
Thanks for doing this! ❤️ 🙏

I am a bit surprised that the left/right semantics could be completely replaced with anchor/active. I'll see if I can come up with a case where we used left/right for something that really meant that.

I'll run with this build some as well, but the changes look good to me. Only thing I'd like to change is to reverse the imports edits.

riotrah committed:
reformat + undo auto import organize

rayat-amperity commented:
> Thanks for doing this! ❤️ 🙏
> 
> I am a bit surprised that the left/right semantics could be completely replaced with anchor/active. I'll see if I can come up with a case where we used left/right for something that really meant that.
> 
> I'll run with this build some as well, but the changes look good to me. Only thing I'd like to change is to reverse the imports edits.

In my multi-cursor branch I added `.start` `.end` `.isReversed` properties to `ModelEditSelection` that are computed at access time (using genius-level Math.max/min calls etc) and update similarly at write time, inspired by the underlying VSCode Selection apis. In some functions, I've replaced references to `.active/.anchor` - or better yet `Math.max/min(anchor, active` - with semantically equivalent `.start` or `.end` references.

If you'd like, those changes may be backported separately from multi-cursor.

PEZ commented on src/cursor-doc/model.ts:
```diff
@@ -1,7 +1,7 @@
-import { Scanner, Token, ScannerState } from './clojure-lexer';
-import { LispTokenCursor } from './token-cursor';
-import { deepEqual as equal } from '../util/object';
 import { isUndefined } from 'lodash';
+import { deepEqual as equal } from '../util/object';
+import { Scanner, ScannerState, Token } from './clojure-lexer';
+import { LispTokenCursor } from './token-cursor';
```
You forgot. 😄 It goes for almost all files in this PR.

riotrah replied:
Aghhh sorry again !!!!

riotrah replied:
Fixed these!

PEZ reviewed (changes requested):
Looks great! Except the imports changes.

PEZ commented:
Is that a question about the changelog I see, @rayat-amperity ? 😄 Yes, we want an entry for changes like this. We usually mark them as `Maintenance: `,  you should find some recent examples. Please refer to the issue you are fixing rather than the PR in the changelog entry. Rationale:

* A changelog entry helps users and us trace when something changed that might have cause some issue someone is seeing.

* Refering to the issue is refering to the problem statement rather than the solution. And since we leave backlinks  to PR by referencing it in the PR text as well as in commit messages, it will be easy to find the PR too.

So, if you haven't included a `Fixes #1607` in a commit message, a good opportunity to do so is when you update the changelog. 😄

riotrah committed:
Undo import auto organization

riotrah committed:
add changelog entry for #1607 selectionLeft/selectionRight API cleanup

riotrah commented:
> Is that a question about the changelog I see, @rayat-amperity ? 😄 Yes, we want an entry for changes like this. We usually mark them as `Maintenance: `, you should find some recent examples. Please refer to the issue you are fixing rather than the PR in the changelog entry. Rationale:
> 
> * A changelog entry helps users and us trace when something changed that might have cause some issue someone is seeing.
> * Refering to the issue is refering to the problem statement rather than the solution. And since we leave backlinks  to PR by referencing it in the PR text as well as in commit messages, it will be easy to find the PR too.
> 
> So, if you haven't included a `Fixes #1607` in a commit message, a good opportunity to do so is when you update the changelog. 😄

Done !

PEZ commented on src/extension-test/unit/cursor-doc/token-cursor-test.ts:
```diff
@@ -1,458 +1,457 @@
       const a = docFromTextNotation('a •|c');
       const b = docFromTextNotation('a| •c');
-      const cursor: LispTokenCursor = a.getTokenCursor(a.selectionLeft);
+      const cursor: LispTokenCursor = a.getTokenCursor(a.selection.anchor);
```
Any particular reason you choose `anchor` over `active` for these? It doesn't matter much, since they are the same, but to me the thing that blinks is the `active` thing. 😄

riotrah replied:
Yeah I agree - good point.

In general most of the time `left` is anchor in the codebase (before this PR I suppose) in selections, and my having used anchor in this case is just an oversight/accident

PEZ reviewed (approved):
Thanks for your work with this, @rayat-amperity !
